zio-http/src/main/scala/zio/http/Headers.scala

```diff
@@ -1,81 +1,151 @@
+ * modify headers.
  *
  * NOTE: Generic operators that are not specific to `Headers` should not be
  * defined here. A better place would be one of the traits extended by
  * `HeaderExtension`.
  */
-final case class Headers(toChunk: Chunk[Header]) extends HeaderExtension[Headers] {
+
+sealed trait Headers extends HeaderExtension[Headers] with Iterable[Header] {
   self =>
+  final def ++(other: Headers): Headers = self.combine(other)
 
-  def ++(other: Headers): Headers = self.combine(other)
+  final def combine(other: Headers): Headers =
+    Headers.Concat(self, other)
 
-  def combine(other: Headers): Headers = Headers(self.toChunk ++ other.toChunk)
+  final def combineIf(cond: Boolean)(other: Headers): Headers =
+    if (cond) self ++ other else self
 
-  def combineIf(cond: Boolean)(other: Headers): Headers = if (cond) Headers(self.toChunk ++ other.toChunk) else self
+  private[http] def encode: HttpHeaders
 
-  override def headers: Headers = self
+  def flatten: Iterable[Header]
 
-  def modify(f: Header => Header): Headers = Headers(toChunk.map(f(_)))
+  override final def headers: Headers = self
 
-  def toList: List[(String, String)] = toChunk.map { case (name, value) => (name.toString, value.toString) }.toList
+  override def iterator: Iterator[Header]
 
-  override def updateHeaders(update: Headers => Headers): Headers = update(self)
+  final def modify(f: Header => Header): Headers = Headers.FromIterable(flatten.map(f))
 
-  def when(cond: Boolean): Headers = if (cond) self else Headers.empty
+  override final def updateHeaders(update: Headers => Headers): Headers = update(self)
 
-  /**
-   * Converts a Headers to [io.netty.handler.codec.http.HttpHeaders]
-   */
-  private[http] def encode: HttpHeaders = {
-    val (exceptions, regularHeaders) = self.toList.span(h => h._1.contains(HeaderNames.setCookie))
-    val combinedHeaders              = regularHeaders
-      .groupBy(_._1)
-      .map { case (key, tuples) =>
-        key -> tuples.map(_._2).mkString(",")
-      }
-    (exceptions ++ combinedHeaders)
-      .foldLeft[HttpHeaders](new DefaultHttpHeaders(true)) { case (headers, entry) =>
-        headers.add(entry._1, entry._2)
-      }
-  }
+  final def when(cond: Boolean): Headers = if (cond) self else Headers.EmptyHeaders
 
 }
 
 object Headers extends HeaderConstructors {
 
-  val empty: Headers   = Headers(Nil)
+  final case class Header private[zio] (key: CharSequence, value: CharSequence)
+      extends Product2[CharSequence, CharSequence]
+      with Headers {
+    self =>
+
+    override def _1: CharSequence = key
+
+    override def _2: CharSequence = value
+
+    override private[http] def encode: HttpHeaders = Headers.encode(self.toList)
+
+    def flatten: Iterable[Header] = Iterable(self)
+
+    override def iterator: Iterator[Header] =
+      Iterator.single(self)
+
+    // Unless we implement it headers.toString results in StackOverflow
+    override def toString(): String = (key, value).toString()
```

Comment:
Looking into if this can be avoided. If it's acceptable as a solution I am ok to leave it like this.